src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -289,68 +342,116 @@ private static void assertVersionMatch() {
         }
     }
 
+    private static void maybeDelete(final Path path) {
+        if (!DebugFlag.PRESERVE_NATIVE_LIBRARIES.isEnabled()) {
+            try {
+                Files.delete(path);
+            } catch (IOException ex) {
+                LOG.warning("Unable to delete native library: " + ex);
+            }
+        }
+    }
+
+    private static void maybeDeletePrivateTempDir(final Path tmpDirectory) {
+        for (String jarResource: JAR_RESOURCES) {
+            String resourceFileName = System.mapLibraryName(jarResource);
+            maybeDelete(tmpDirectory.resolve(resourceFileName));
+        }
+        maybeDelete(tmpDirectory);
+    }
+
+    private static Path writeJarResourceToTemporaryFile(Path tempDirectory, final String resourceFileName) throws IOException {
+        final Path tempResourceFilePath = tempDirectory.resolve(resourceFileName);
+
+        // Create a new temporary file. If one already exists this will throw FileAlreadyExistsException
```

Comment:
Please expand this comment to explain why it is important that we create first:
We need to ensure that we create the file and set permissions atomically so that an adversary cannot put a symlink or file here which would cause us to write (or read) to an arbitrary attacker controlled location.